### DISCUSSION TITLE
Github action use a secret depending the context

### DISCUSSION THREAD

**Discussion by raphaelauv** [community/community#25260] [Actions]
<p>I have two secrets , one for PREPROD and one for PROD , and my github action should use the secret depending if it’s a run of preprod or prod</p>
<pre><code class="lang-yaml">
      - name: Sets env vars for PREPROD
        if: ${{ github.event.inputs.environment == 'preprod' }}
        run: |
          echo "GKE_CLUSTER=gke-data-preprod" &gt;&gt; $GITHUB_ENV
          echo "GKE_ZONE=europe-west1" &gt;&gt; $GITHUB_ENV
          echo "GCP_SECRET=secrets.GCP_SERVICE_ACCOUNT_KEY_PREPROD" &gt;&gt; $GITHUB_ENV

      - name: Sets env vars for PROD
        if: ${{ github.event.inputs.environment == 'prod' }}
        run: |
          echo "GKE_CLUSTER=gke-data-prod" &gt;&gt; $GITHUB_ENV
          echo "GKE_ZONE=europe-west1" &gt;&gt; $GITHUB_ENV
          echo "GCP_SECRET=secrets.GCP_SERVICE_ACCOUNT_KEY_PROD" &gt;&gt; $GITHUB_ENV

      - id: "auth"
        uses: "google-github-actions/auth@v0"
        with:
          credentials_json: "${{ ${{ env.GCP_SECRET }} }}"
</code></pre>
<p>but this do not work , it fail with this error</p>
<pre><code class="lang-auto">The workflow is not valid .... Unexpected symbol: '${{'. Located at position 1 within expression: ${{ env.GCP_SECRET
</code></pre>
<p>Also I can’t do</p>
<pre><code class="lang-yaml">      - name: Sets env vars for PREPROD
        if: ${{ github.event.inputs.environment == 'preprod' }}
        run: |
          echo "GKE_CLUSTER=gke-data-preprod" &gt;&gt; $GITHUB_ENV
          echo "GKE_ZONE=europe-west1" &gt;&gt; $GITHUB_ENV
          echo "GCP_SECRET=${{ secrets.GCP_SERVICE_ACCOUNT_KEY_PREPROD }}" &gt;&gt; $GITHUB_ENV

      - name: Sets env vars for PROD
        if: ${{ github.event.inputs.environment == 'prod' }}
        run: |
          echo "GKE_CLUSTER=gke-data-prod" &gt;&gt; $GITHUB_ENV
          echo "GKE_ZONE=europe-west1" &gt;&gt; $GITHUB_ENV
          echo "GCP_SECRET=${{ secrets.GCP_SERVICE_ACCOUNT_KEY_PROD }}" &gt;&gt; $GITHUB_ENV

      - id: "auth"
        uses: "google-github-actions/auth@v0"
        with:
          credentials_json: "${{ env.GCP_SECRET }}"
</code></pre>
<p>Because the secrets.GCP_SERVICE_ACCOUNT_KEY_PROD is a multiline secret and it fail with</p>
<pre><code class="lang-auto">echo "GCP_SECRET=***
    ***
    ***
    ***
    ***
    ***
    ***
    ***
    ***
    ***
    ***
  ***" &gt;&gt; $GITHUB_ENV
  shell: /usr/bin/bash -e ***0***
Error: Unable to process file command 'env' successfully.
Error: Invalid environment variable format '  type: service_account,'

</code></pre>

**Comment by jsoref** [comment, marked as answer]
<p>There are lots of things you could do. And lots of explanations.</p>
<p>But, why not do this instead?</p>
<pre><code class="lang-auto">      - name: Sets env vars for PREPROD
        if: ${{ github.event.inputs.environment == 'preprod' }}
        run: |
          echo "GKE_CLUSTER=gke-data-preprod" &gt;&gt; $GITHUB_ENV
          echo "GKE_ZONE=europe-west1" &gt;&gt; $GITHUB_ENV

      - id: "auth-preprod"
        if: ${{ github.event.inputs.environment == 'preprod' }}
        uses: "google-github-actions/auth@v0"
        with:
          credentials_json: "${{ secrets.GCP_SERVICE_ACCOUNT_KEY_PREPROD" }}"

      - name: Sets env vars for PROD
        if: ${{ github.event.inputs.environment == 'prod' }}
        run: |
          echo "GKE_CLUSTER=gke-data-prod" &gt;&gt; $GITHUB_ENV
          echo "GKE_ZONE=europe-west1" &gt;&gt; $GITHUB_ENV

      - id: "auth-prod"
        if: ${{ github.event.inputs.environment == 'prod' }}
        uses: "google-github-actions/auth@v0"
        with:
          credentials_json: "${{ secrets.GCP_SERVICE_ACCOUNT_KEY_PROD }}"
</code></pre>

**Comment by raphaelauv** [comment]
<p>Thanks , your solution is great ! I didn’t knew we could put a IF condition before a USES.</p>

**Comment by jsoref** [comment]
<p>More precisely, you’re putting an <code>if</code> into a <code>steps</code> element, and <code>uses</code> happens to be a sibling:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsif">
  <header class="source">
      <img src="https://docs.github.com/assets/cb-803/images/site/favicon.svg" class="site-icon" width="16" height="16">

      <a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsif" target="_blank" rel="noopener">GitHub Docs</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/2119212/181089941-426d08aa-0148-4c17-b710-c416e8fcc201.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsif" target="_blank" rel="noopener">Workflow syntax for GitHub Actions - GitHub Docs</a></h3>

  <p>A workflow is a configurable automated process made up of one or more jobs. You must create a YAML file to define your workflow configuration.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
